src/client.ts

```diff
@@ -111,12 +111,21 @@ export class OAuth2Client {
   /**
    * Retrieves an OAuth2 token using the client_credentials grant.
    */
-  async clientCredentials(params?: {scope?: string[]}): Promise<OAuth2Token> {
+  async clientCredentials(params?: { scope?: string[]; resource?: string[]; extraParams?: Record<string, string> }): Promise<OAuth2Token> {
 
-    const body:ClientCredentialsRequest = {
+    const disallowed = ['client_id', 'client_secret', 'scope', 'resource'];
+
+    if (params?.extraParams && Object.keys(params.extraParams).filter((key) => disallowed.includes(key)).length > 0) {
+      throw new Error(`The following extraParams are disallowed: '${disallowed.join("', '")}'`);
+    }
+
+    const body: ClientCredentialsRequest = {
       grant_type: 'client_credentials',
       scope: params?.scope?.join(' '),
+      resource: params?.resource?.join(' '),
```

Comment:
So the way the `resource` is encoded (unlike) `scope` is not space-separated. Instead the parameter is encoded multiple times.

Honestly a pretty unfortunate choice, because most languages have libraries to encode/decode these bodies (and urls) but many map to an object.

So either this should be fixed to support multiple `resource` parameters, or we leave encoding resource to a future PR.

Reply:
Hmm, I suppose that also means values in `extraParams` _may_ need to be able to support multiple values too?

Do you think it'd be a better approach to instead utilize `UrlSearchParams` which allows us to `.append(key, value)` and then `.toString()` the whole object URL encoded...?

Reply:
Something like this?

```ts
/**
 * Retrieves an OAuth2 token using the client_credentials grant.
 */
async clientCredentials(
  params: {
    scope?: string[];
    resource?: string | string[];
    extraParams?: Record<string, string | string[]>;
  } = {}
): Promise<OAuth2Token> {
  const { scope, resource, extraParams } = params;

  const disallowed = ["client_id", "client_secret", "grant_type", "scope", "resource"];

  if (extraParams && Object.keys(extraParams).filter((key) => disallowed.includes(key)).length > 0) {
    throw new Error(`The following extraParams are disallowed: '${disallowed.join("', '")}'`);
  }

  const searchParams = new URLSearchParams();

  searchParams.append("grant_type", "client_credentials");

  if (scope) {
    // The value of the scope parameter is expressed as a list of
    // space-delimited, case-sensitive strings rather than multiple parameters.
    // See: https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-08#section-3.2.2.1
    searchParams.append("scope", scope.join(" "));
  }

  if (resource) {
    if (Array.isArray(resource)) {
      resource.forEach((value) => searchParams.append("resource", value));
    } else {
      searchParams.append("resource", resource);
    }
  }

  if (extraParams) {
    Object.entries(extraParams).forEach(([key, value]) => {
      if (Array.isArray(value)) {
        value.forEach((v) => searchParams.append(key, v));
        return;
      }

      searchParams.append(key, value);
    });
  }

  if (!this.settings.clientSecret) {
    throw new Error("A clientSecret must be provided to use client_credentials");
  }

  return tokenResponseToOAuth2Token(this.request("tokenEndpoint", searchParams));
}
```

Of course, I've omitted changing the `request` method here to support this. While the code gets a bit more complex with this - the benefit would be that in the request method we just call `searchParams.toString()` and set the body to the resulting query string

This may be better suited to a new PR however with additional unit tests in place to ensure that existing functionality isn't broken (ie; no major version required 😉)

Reply:
I'd say lets cross that bridge when we get there, a simple `Record<string, string>` type is good enough for now (and hopefully rare enough)